include/nbl/video/IGPUCommandBuffer.h

```diff
@@ -227,6 +232,93 @@ class IGPUCommandBuffer :
             for (uint32_t i = _first + 1u; i < IGPUPipelineLayout::DESCRIPTOR_SET_COUNT; ++i)
                 _destPplnLayouts[i] = nullptr;
     }
+
+    template <typename Cmd, typename... Args>
+    Cmd* emplace(Args&&... args)
+    {
+        if (m_segmentListTail == nullptr)
+        {
+            void* cmdSegmentMem = m_cmdpool->m_commandSegmentPool.allocate(IGPUCommandPool::COMMAND_SEGMENT_SIZE, alignof(IGPUCommandPool::CommandSegment));
+            if (!cmdSegmentMem)
+                return nullptr;
+
+            m_segmentListTail = new (cmdSegmentMem) IGPUCommandPool::CommandSegment;
+            m_segmentListHeadItr.m_segment = m_segmentListTail;
+        }
+
+        // Cmd* cmd = m_segmentListTail->allocate<Cmd, Args...>(args...);
+        Cmd* cmd = m_segmentListTail->allocate<Cmd>(std::forward<Args>(args)...);
+        if (!cmd)
+        {
+            void* nextSegmentMem = m_cmdpool->m_commandSegmentPool.allocate(IGPUCommandPool::COMMAND_SEGMENT_SIZE, alignof(IGPUCommandPool::CommandSegment));
+            if (nextSegmentMem == nullptr)
+                return nullptr;
+
+            IGPUCommandPool::CommandSegment* nextSegment = new (nextSegmentMem) IGPUCommandPool::CommandSegment;
+
+            // cmd = m_segmentListTail->allocate<Cmd, Args...>(args...);
+            cmd = m_segmentListTail->allocate<Cmd>(std::forward<Args>(args)...);
+            if (!cmd)
+                return nullptr;
+
+            m_segmentListTail->params.m_next = nextSegment;
+            m_segmentListTail = m_segmentListTail->params.m_next;
+        }
+
+        if (m_segmentListHeadItr.m_cmd == nullptr)
+            m_segmentListHeadItr.m_cmd = cmd;
+
+        return cmd;
+    }
+
+    IGPUCommandPool::CommandSegment::Iterator m_segmentListHeadItr = {};
```

Comment:
if your command segment can have a `begin()` and `end()` method, then all you need is just
```cpp
IGPUCommandPool::CommandSegment* m_segmentListHead = nullptr;
```

Reply:
> also I think these should be private (derived command buffer shouldn't be able to mess with the segmented list pointers)

Yep, made them `private`.

Reply:
but you can still make it just a `IGPUCommandPool::CommandSegment*` instead of the iterator